### PR TITLE
[sui-adapter] Move execution_mode and programmable_transaction/types

### PR DESCRIPTION
**Pull request by amnn**
## Description

...to `sui-types`, as another part of the interface between the execution layer and its callers.  The following changes were made to the interface to make this work:

- `ExecutionMode` functions that previously accepted a `&mut ExecutionContext` now accept a `&impl TypeTagResolver`, as that is all this parameter was used for (for dev inspect transactions). This also simplifies the constraints on generic type parameters in this module.
- `ObjectValue` no longer has constructors from `MoveObject` etc, because it has been moved to `sui-types` but these constructors are implemented in terms of the execution layer. They have been replaced by private free functions within `sui-adapter`.

## Test Plan

```
$ cargo simtest
$ env SUI_SKIP_SIMTESTS=1 cargo nextest run
```

### CODE REVIEW

File: crates/sui-types/src/execution.rs [hunk 1]
```diff
@@ -183,69 +177,6 @@ impl Value {
 }
 
 impl ObjectValue {
-    pub fn new<'vm, 'state, S: StorageView>(
```
Comment (amnn) [inline crates/sui-types/src/execution.rs:186]: Turned into free functions in `programmable_transactions/context.rs`

File: crates/sui-types/src/execution_mode.rs [hunk 2]
```diff
@@ -32,23 +33,19 @@ pub trait ExecutionMode {
 
     fn empty_results() -> Self::ExecutionResults;
 
-    fn add_argument_update<'vm, 'state, 'a, S: StorageView>(
-        context: &mut ExecutionContext<'vm, 'state, 'a, S>,
```
Comment (amnn) [inline crates/sui-types/src/execution_mode.rs:36]: `ExecutionContext` is part of the execution layer, so we can't depend on it here -- it's abstracted behind a trait.

File: crates/sui-types/src/error.rs [hunk 3]
```diff
@@ -73,7 +74,7 @@ macro_rules! invariant_violation {
 
 #[macro_export]
 macro_rules! assert_invariant {
-    ($cond:expr, $($args:expr),*) => {{
+    ($cond:expr, $($args:expr),* $(,)?) => {{
```
Comment (rvantonder) [inline crates/sui-types/src/error.rs:77]: Looks like the `assert_invariant` call changed below doesn't have a trailing comma, so think this change isn't needed, but might be good practice anyway? Am not a macro expert :)

**Review by rvantonder — APPROVED** [review]
Without being intimately familiar with all execution layer things yet, this looks clean and fairly benign, so approving to unblock!

File: crates/sui-types/src/error.rs (hunk 3 above)
Comment (amnn) [reply crates/sui-types/src/error.rs:77]: Ah yes!  I'm not sure what combination of formatters and my own habits got me here, but you're right, this is not strictly necessary anymore, but is good practice anyway.

**Pull request merged**